File: csharp/ShoppingCart.cs
using System;
using System.Collections.Generic;

namespace supermarket
{
    public class ShoppingCart
    {

        private List<ProductQuantity> items = new List<ProductQuantity>();
        Dictionary<Product, double> productQuantities = new Dictionary<Product, double>();


        public List<ProductQuantity> GetItems()
        {
            return new List<ProductQuantity>(items);
        }

        public void AddItem(Product product)
        {
            this.AddItemQuantity(product, 1.0);
        }


        public void AddItemQuantity(Product product, double quantity)
        {
            items.Add(new ProductQuantity(product, quantity));
            if (productQuantities.ContainsKey(product))
            {
                var newAmount = productQuantities[product] + quantity;
                productQuantities[product] = newAmount;
            }
            else
            {
                productQuantities.Add(product, quantity);
            }
        }

        public void HandleOffers(Receipt receipt, Dictionary<Product, Offer> offers, SupermarketCatalog catalog)
        {
            foreach (Product p in productQuantities.Keys)
            {
                double quantity = productQuantities[p];
                if (offers.ContainsKey(p))
                {
                    Offer offer = offers[p];
                    double unitPrice = catalog.GetUnitPrice(p);
                    int quantityAsInt = (int)quantity;
                    Discount discount = null;
                    int x = 1;
                    if (offer.OfferType == SpecialOfferType.ThreeForTwo)
                    {
                        x = 3;

                    }
                    else if (offer.OfferType == SpecialOfferType.TwoForAmount)
                    {
                        x = 2;
                        if (quantityAsInt >= 2)
                        {
                            double total = offer.Argument * quantityAsInt / x + quantityAsInt % 2 * unitPrice;
                            double discountN = unitPrice * quantity - total;
                            discount = new Discount(p, "2 for " + offer.Argument, discountN);
                        }

                    }
                    if (offer.OfferType == SpecialOfferType.FiveForAmount)
                    {
                        x = 5;
                    }
                    int numberOfXs = quantityAsInt / x;
                    if (offer.OfferType == SpecialOfferType.ThreeForTwo && quantityAsInt > 2)
                    {
                        double discountAmount = quantity * unitPrice - ((numberOfXs * 2 * unitPrice) + quantityAsInt % 3 * unitPrice);
                        discount = new Discount(p, "3 for 2", discountAmount);
                    }
                    if (offer.OfferType == SpecialOfferType.TenPercentDiscount)
                    {
                        discount = new Discount(p, offer.Argument + "% off", quantity * unitPrice * offer.Argument / 100.0);
                    }
                    if (offer.OfferType == SpecialOfferType.FiveForAmount && quantityAsInt >= 5)
                    {
                        double discountTotal = unitPrice * quantity - (offer.Argument * numberOfXs + quantityAsInt % 5 * unitPrice);
                        discount = new Discount(p, x + " for " + offer.Argument, discountTotal);
                    }
                    if (discount != null)
                        receipt.addDiscount(discount);
                }

            }
        }
    }

}

File: cpp/src/ShoppingCart.cpp
#include "ShoppingCart.h"

void addItemQuantity(const Product& product, double quantity);

std::vector<ProductQuantity> ShoppingCart::getItems() const {
    return items;
}

std::map<Product, double> ShoppingCart::getProductQuantities() const {
    return productQuantities;
}

void ShoppingCart::addItem(const Product& product) {
    addItemQuantity(product, 1.0);
}

void ShoppingCart::addItemQuantity(const Product& product, double quantity) {
    items.emplace_back(product, quantity);
    if (productQuantities.find(product) != productQuantities.end()) {
        productQuantities[product] += quantity;
    } else {
        productQuantities[product] = quantity;
    }
}

void ShoppingCart::handleOffers(Receipt& receipt, std::map<Product, Offer> offers, SupermarketCatalog* catalog) {
    for (const auto& productQuantity : productQuantities) {
        Product product = productQuantity.first;
        double quantity = productQuantity.second;
        if (offers.find(product) != offers.end()) {
            auto offer = offers[product];
            double unitPrice = catalog->getUnitPrice(product);
            int quantityAsInt = (int) quantity;
            Discount* discount = nullptr;
            int x = 1;

            if (offer.getOfferType() == SpecialOfferType::ThreeForTwo) {
                x = 3;
            } else if (offer.getOfferType() == SpecialOfferType::TwoForAmount) {
                x = 2;
                if (quantityAsInt >= 2) {
                    double total = offer.getArgument() * (quantityAsInt / x) + quantityAsInt % 2 * unitPrice;
                    double discountN = unitPrice * quantity - total;
                    discount = new Discount("2 for " + std::to_string(offer.getArgument()), -discountN, product);
                }
            } if (offer.getOfferType() == SpecialOfferType::FiveForAmount) {
                x = 5;
            }
            int numberOfXs = quantityAsInt / x;
            if (offer.getOfferType() == SpecialOfferType::ThreeForTwo && quantityAsInt > 2) {
                double discountAmount = quantity * unitPrice - ((numberOfXs * 2 * unitPrice) + quantityAsInt % 3 * unitPrice);
                discount = new Discount("3 for 2", -discountAmount, product);
            }
            if (offer.getOfferType() == SpecialOfferType::TenPercentDiscount) {
                discount = new Discount(std::to_string(offer.getArgument()) + "% off", -quantity * unitPrice * offer.getArgument() / 100.0, product);
            }
            if (offer.getOfferType() == SpecialOfferType::FiveForAmount && quantityAsInt >= 5) {
                double discountTotal = unitPrice * quantity - (offer.getArgument() * numberOfXs + quantityAsInt % 5 * unitPrice);
                discount = new Discount(std::to_string(x) + " for " + std::to_string(offer.getArgument()), -discountTotal, product);
            }
            if (discount != nullptr)
                receipt.addDiscount(*discount);
        }
    }
}